src/contracts/AbstractARM.sol

```diff
@@ -0,0 +1,714 @@
+    function _transferAssetFrom(address asset, address from, address to, uint256 amount) internal virtual {
+        IERC20(asset).transferFrom(from, to, amount);
+    }
+
+    function _swapExactTokensForTokens(IERC20 inToken, IERC20 outToken, uint256 amountIn, address to)
+        internal
+        virtual
+        returns (uint256 amountOut)
+    {
+        uint256 price;
+        if (inToken == token0) {
+            require(outToken == token1, "ARM: Invalid out token");
+            price = traderate0;
+        } else if (inToken == token1) {
+            require(outToken == token0, "ARM: Invalid out token");
+            price = traderate1;
+        } else {
+            revert("ARM: Invalid in token");
+        }
+        amountOut = amountIn * price / PRICE_SCALE;
+
+        // Transfer the input tokens from the caller to this ARM contract
+        _transferAssetFrom(address(inToken), msg.sender, address(this), amountIn);
+
+        // Transfer the output tokens to the recipient
+        _transferAsset(address(outToken), to, amountOut);
+    }
+
+    function _swapTokensForExactTokens(IERC20 inToken, IERC20 outToken, uint256 amountOut, address to)
+        internal
+        virtual
+        returns (uint256 amountIn)
+    {
+        uint256 price;
+        if (inToken == token0) {
+            require(outToken == token1, "ARM: Invalid out token");
+            price = traderate0;
+        } else if (inToken == token1) {
+            require(outToken == token0, "ARM: Invalid out token");
+            price = traderate1;
+        } else {
+            revert("ARM: Invalid in token");
+        }
+        amountIn = ((amountOut * PRICE_SCALE) / price) + 1; // +1 to always round in our favor
+
+        // Transfer the input tokens from the caller to this ARM contract
+        _transferAssetFrom(address(inToken), msg.sender, address(this), amountIn);
+
+        // Transfer the output tokens to the recipient
+        _transferAsset(address(outToken), to, amountOut);
+    }
+
+    /**
+     * @notice Set exchange rates from an operator account from the ARM's perspective.
+     * If token 0 is WETH and token 1 is stETH, then both prices will be set using the stETH/WETH price.
+     * @param buyT1 The price the ARM buys Token 1 (stETH) from the Trader, denominated in Token 0 (WETH), scaled to 36 decimals.
+     * From the Trader's perspective, this is the sell price.
+     * @param sellT1 The price the ARM sells Token 1 (stETH) to the Trader, denominated in Token 0 (WETH), scaled to 36 decimals.
+     * From the Trader's perspective, this is the buy price.
+     */
+    function setPrices(uint256 buyT1, uint256 sellT1) external onlyOperatorOrOwner {
+        // Ensure buy price is always below past sell prices
+        require(sellT1 >= crossPrice, "ARM: sell price too low");
+        require(buyT1 < crossPrice, "ARM: buy price too high");
+
+        traderate0 = PRICE_SCALE * PRICE_SCALE / sellT1; // base (t0) -> token (t1);
```

Comment:
🔴 So probably the single most important invariant in the system is no local price crosses. If a round trip pair of trades loses us money, then someone can loop that to drain us. This code does ensure that raw incoming prices stay on other sides of the price cross fence.

However, the traderates are the actual number used in calculating trades. Roughly speaking `amountIn * traderate == amountOut`. This means that high trade rates are good for us, low trade rates are bad. This line rounds the wrong way, rounding against the protocol in calc the trade rate.

It's possible there's no way to get it to cross and or gains impossible,  but I like the rounding going the right way. :D

Reply:
A high stETH/WETH sell price is good for us, yes. That's when the ARM input token is WETH and output token is stETH.
But `traderate0` inverts the stETH/WETH sell price so we want a low `traderate0` so the truncation works in our favour here. That is, we want to convert the incoming WETH to a lower amount of stETH going out.

```Solidity
traderate0 = PRICE_SCALE * PRICE_SCALE / sellT1; // quote (t0) -> base (t1); eg WETH -> stETH
```

I've added the token examples to make it clearer which was the stETH and WETH is going.

Reply:
You are correct. Down is good.